python/cudf/cudf/_lib/pylibcudf/io/types.pyx

```diff
@@ -84,6 +86,13 @@ cdef class SourceInfo:
 
             self.c_obj = move(source_info(c_files))
             return
+        elif isinstance(sources[0], Datasource):
+            for csrc in sources:
```

Comment:
Let's go ahead and insert the deprecation warning as part of this PR, either here or in the NativeFileDatasource constructor.